### PR TITLE
fix(plugin): get owner node id from created resource

### LINKED ISSUE #541: Volume From Snapshot Fails With Custom Node Ids
**What steps did you take and what happened:**
[A clear and concise description of what the bug is, and what commands you ran.]

1. Label node(s) with unique value for `openebs.io/nodeid`
2. Deploy zfs components
3. Create a pvc with zfs storage class
4. Identify volume created successfully
5. Create a snapshot and ensure the appropriate zfs and k8s resources exist
6. Create a new pvc defined to be created from snapshot
7. Notice zfs volume created and after 10 minutes goes to pending
8. Logs will indicate unable to find node with nodeid value from the owning node

**What did you expect to happen:**
It should utilize the correct ids when identifying the node and create the volume from snapshot successfully.

**Anything else you would like to add:**
We pre-label (before starting zfs components) our nodes with unique identifiers for `openebs.io/nodeid` and these identifiers differ from the node name.  

With this configuration, I am able to create new ZFS volumes without issue. However, when you attempt to create a zfs volume from a snapshot definition it fails. 

In the logs I can see `failed to get the node {VALUE_OF_OPENEBS_NODEID}` which highlights the issue. This `GetNodeID` method in `volume.go` only works when it is passed the nodename.

In the case of creating a new volume `GetNodeID` is called with the nodename but when it is from a volume clone or a snapshot it gets called with the value of `openebs.io/nodeid`. This is because this block here in `driver/controller.go` in the `CreateVolume` method:

```
if contentSource != nil && contentSource.GetSnapshot() != nil {
		snapshotID := contentSource.GetSnapshot().GetSnapshotId()

		selected, err = CreateSnapClone(ctx, req, snapshotID)
	} else if contentSource != nil && contentSource.GetVolume() != nil {
		srcVol := contentSource.GetVolume().GetVolumeId()
		selected, err = CreateVolClone(ctx, req, srcVol)
	} else {
		selected, err = CreateZFSVolume(ctx, req)
	}
```

`CreateSnapClone` --> returns spec.OwnerNodeId
`CreateVolClone` --> returns spec.OwnerNodeId
`CreateZFSVolume` --> returns nodename loop variable

**Environment:**
- LocalPV-ZFS version: 4.0.1
- Kubernetes version (use `kubectl version`): 1.29
- Kubernetes installer & version: GKE
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jnels124**
## Pull Request template

Please, go through these steps before you submit a PR.

**Why is this PR required? What issue does it fix?**:
* Allows restoring from snapshot when nodeid differs from the node name

**What this PR does?**:
* gets the owner node id from the created resource in all cases

**Does this PR require any upgrade changes?**:
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:
1. Deployed with zfs nodes and custom nodeid
2. Ingested data
3. Created a snapshot from disks
4. Restored from snapshots
5. verified data present

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [x] Fixes #541
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/website is used to track them:

**PLEASE REMOVE BELOW INFORMATION BEFORE SUBMITTING**

The PR title message must follow convention:
   `<type>(<scope>): <subject>`.

Where: <br />
- `type` is defining if release will be triggering after merging submitted changes, details in [CONTRIBUTING.md](../CONTRIBUTING.md).
    Most common types are:
    - `feat`      - for new features, not a new feature for build script
    - `fix`       - for bug fixes or improvements, not a fix for build script
    - `chore`     - changes not related to production code
    - `docs`      - changes related to documentation
    - `style`     - formatting, missing semi colons, linting fix etc; no significant production code changes
    - `test`      - adding missing tests, refactoring tests; no production code change
    - `refactor`  - refactoring production code, eg. renaming a variable or function name, there should not be any significant production code changes

- `scope` is a single word that best describes where the changes fit.
  Most common scopes are like:
    - data engine (`localpv`, `jiva`, `cstor`)
    - feature (`provisioning`, `backup`, `restore`, `exporter`)
    - code component (`api`, `webhook`, `cast`, `upgrade`)
    - test (`tests`, `bdd`)
    - chores (`version`, `build`, `log`, `travis`)

- `subject` is a single line brief description of the changes made in the pull request.

### CODE REVIEW

**Comment by niladrih** [comment]
@jnels124 -- do you think you could help me with a test case for this nodeId changes that we've added recently?

**Comment by jnels124** [comment]
> @jnels124 -- do you think you could help me with a test case for this nodeId changes that we've added recently?

Yes absolutely.

**Comment by niladrih** [comment]
@jnels124 -- Thank you! We use the ginkgo test framework. The tests sit inside the 'tests' directory. They run on a single-node minikube ('none' driver) kubernetes cluster in our CI. Let me know if you're blocked with any of this. Again, thank you!

**Comment by jnels124** [comment]
@niladrih Please review this approach. This of course doubles the test execution time but the test 
cases present make sense to run with this configuration. Note, the existing tests don't verify much in many cases. Will create a follow up ticket to address that separately.

**Comment by Abhinandan-Purkait** [comment]
> @niladrih Please review this approach. This of course doubles the test execution time but the test cases present make sense to run with this configuration. Note, the existing tests don't verify much in many cases. Will create a follow up ticket to address that separately.

cc @niladrih

**Review by Abhinandan-Purkait — COMMENTED** [review]
The code changes look good to me. However running the ci tests two times looks a bit of an overkill to me, maybe this should be an e2e use case rather than a ci test. Doubling ci affects the PR, PR merge... all workflows.

**Review by niladrih — APPROVED** [review]
Thank you for the work that you've put in here @jnels124! I'm inclined to say that we should both of the sets of tests in parallel. However, the tests are taking 30 min, which isn't too bad and it is testing the core features. I'm okay with this tbh. LGTM

**Comment by tiagolobocastro** [comment]
> The code changes look good to me. However running the ci tests two times looks a bit of an overkill to me, maybe this should be an e2e use case rather than a ci test. Doubling ci affects the PR, PR merge... all workflows.

Yeah I'd also suggest running a smaller number of tests with custom node id, at least on PR being raised. Maybe we can have extended tests running at a separate cadence.

**Review by niladrih — COMMENTED** [review]
Looking at this again... this approach might scale badly as newer tests are added, as they'd be run twice.

It makes sense to run a sanity test set with volume provision/de-provision and snapshot creation/deletion as a compromise.

It might be a good idea to add a copy of the test set at https://github.com/openebs/zfs-localpv/blob/develop/tests/provision_test.go in a new file. The `prepareCustomNodeIdEnv` would be in code, and it would be run as the very first `By` (before `By("####### Creating the storage class : " + fstype + " #######")` or `By("Creating default storage class", createStorageClass)`) in `fsVolCreationTest()` and `blockVolCreationTest()` functions. This way what changes is, if there are new tests added they are not run twice, and nodeId is tested for the big workflows. WDYT about this approach @jnels124?

**Comment by jnels124** [comment]
> Looking at this again... this approach might scale badly as newer tests are added, as they'd be run twice.
> 
> It makes sense to run a sanity test set with volume provision/de-provision and snapshot creation/deletion as a compromise.
> 
> It might be a good idea to add a copy of the test set at https://github.com/openebs/zfs-localpv/blob/develop/tests/provision_test.go in a new file. The `prepareCustomNodeIdEnv` would be in code, and it would be run as the very first `By` (before `By("####### Creating the storage class : " + fstype + " #######")` or `By("Creating default storage class", createStorageClass)`) in `fsVolCreationTest()` and `blockVolCreationTest()` functions. This way what changes is, if there are new tests added they are not run twice, and nodeId is tested for the big workflows. WDYT about this approach @jnels124?

That makes sense to me I will get that in. Will let you know if I encounter difficulties around the environment config. I should be able to update node labels and delete pods to configure the environment appropriately. Will also need to ensure the test resets the environment back to the original state. This also means, this test can not be ran in parallel with other tests.

**Comment by jnels124** [comment]
@niladrih I took a slightly different approach but accomplishes the goal of running only specific tests after the environment has been configured with a custom node id. I went with this approach because there is not an existing client (and other things) in the test set up to allow modification of k8s nodes. Also, this approach allows more flexibility for future testing

**Comment by Abhinandan-Purkait** [comment]
@jnels124 Can you please rebase the PR branch, a lot of changes went in on the tests and ci. I see we are bumping go version as well, just wanted to double check with @niladrih @tiagolobocastro @avishnu about this.

File: go.mod [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
 
-go 1.19
+go 1.20
 
 require (
 	github.com/container-storage-interface/spec v1.8.0
 	github.com/kubernetes-csi/csi-lib-utils v0.9.0
-	github.com/onsi/ginkgo v1.16.5
-	github.com/onsi/gomega v1.27.7
+	github.com/onsi/ginkgo/v2 v2.19.0
```
Comment (niladrih) [inline go.mod]:
```suggestion
	github.com/onsi/ginkgo/v2 v2.19.1
```

**Review by niladrih — COMMENTED** [review]
Oh, I missed the go version bump. The go bump needs to go into this Dockerfile as well https://github.com/openebs/zfs-localpv/blob/develop/buildscripts/zfs-driver/zfs-driver.Dockerfile.

File: go.mod [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
+	github.com/onsi/gomega v1.33.1
```
Comment (niladrih) [inline go.mod]: Good idea to bump gomega as well
```suggestion
	github.com/onsi/gomega v1.34.1
```

File: .github/workflows/pull_request.yml [hunk 3]
```diff
@@ -109,10 +109,10 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v4
 
-      - name: Set up Go 1.19
+      - name: Set up Go 1.20
```
Comment (Abhinandan-Purkait) [inline .github/workflows/pull_request.yml:112]: Can you please update the version on line no 84 as well?

Comment (jnels124) [reply .github/workflows/pull_request.yml:112]: apologies will get that done

**Review by Abhinandan-Purkait — APPROVED** [review]

**Review by niladrih — APPROVED** [review]
LGTM, thanks.

**Comment by niladrih** [comment]
@jnels124 thank you for making the change! You'd have to rebase your PR branch. I see commits from @sinhaashish and @Abhinandan-Purkait on your branch. This is likely because of a merge, instead of a rebase. I think there might be a git rebase option `--rebase-merges`.

What we're trying to do is to get your commits to sit on top. Let me know if you need help with this.

**Comment by jnels124** [comment]
@niladrih I have rebased the commits and pr should only contain my feature commits. thanks

**Comment by Abhinandan-Purkait** [comment]
@jnels124 Thanks, are we good to merge now?

**Comment by jnels124** [comment]
@Abhinandan-Purkait yes everything is complete and good to merge

**Comment by Abhinandan-Purkait** [comment]
@jnels124 Thanks. You can test the changes using the `2.7.0-develop` helm chart, till the next release is cut.

**Pull request merged**